core/database/migration/migration.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package migration
+
+import (
+	"context"
+
+	"github.com/juju/description/v4"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+)
+
+// Operation is a single step in a migration.
+type Operation interface {
+	// Setup is called before the operation is executed. It should return an
+	// error if the operation cannot be performed.
+	Setup(database.DBGetter) error
+
+	// Execute is called to perform the operation. It should return an error
+	// if the operation fails.
+	Execute(context.Context, description.Model) error
+
+	// Rollback is called if the operation fails. It should attempt to undo
+	// any changes made by the operation. This is best effort, and may not
+	// always be possible.
+	Rollback(context.Context) error
+}
+
+// Hook is a callback that is called after the operation is executed.
+type Hook func(Operation) error
+
+// Migration is a collection of operations that can be performed as a single
+// unit. This is not atomic, but it does allow for a rollback of the entire
+// migration if any operation fails.
+type Migration struct {
+	operations []Operation
+	hook       Hook
+}
+
+// New creates a new migration coordinator with the given operations.
+func New(operations ...Operation) *Migration {
+	return &Migration{
+		operations: operations,
+		hook:       omitHook,
```

Comment:
what's (or what will be) the point of this hook?

Reply:
It might not be needed, but I might also offer more arguments i.e. the txn runner so that you can perform additional work after a migration. For example, do a request etc.